src/experimental/slam_toolbox_lifelong.cpp

```diff
@@ -76,8 +81,683 @@ LifelongSlamToolbox::LifelongSlamToolbox(rclcpp::NodeOptions options)
+}
+
+float LifelongSlamToolbox::calculateMapMutualInformation()
+{
+  /*
+    To calculate map mutual information, this is the summation 
+    of all cells mutual information
+  */
+  float sum = 0.0f;
+  for (int i = 0; i < m_num_cells; ++i)
+  {
+    for (int j = 0; j < m_num_cells; ++j)
+    {
+      sum += m_mutual_grid[i][j];
+    }
+  }
+  return sum;
+}
+
+void LifelongSlamToolbox::calculateLimits(
+  std::vector<float> & initial_x, std::vector<float> & initial_y, std::vector<float> & final_x, std::vector<float> & final_y, 
+  float & limit_x, float & limit_y, float & min_x, float & max_x, float & min_y, float & max_y, std::vector<int> & robot_grid_pos, 
+  std::vector<int> & final_grid_pos)
+{
+  /*
+    To calculate grid grid limits for intersection
+  */
+  if (final_grid_pos[0] < robot_grid_pos[0] && final_grid_pos[1] >= robot_grid_pos[1])
+  {
+    // X greater and Y greater. WRO final points
+    final_x[0] = limit_x + m_resolution;
+    final_x[2] = limit_x + m_resolution;
+
+    min_y = limit_y;
+    max_y = limit_y + m_resolution;
+  }
+
+  if (final_grid_pos[0] >= robot_grid_pos[0] && final_grid_pos[1] < robot_grid_pos[1])
+  {
+    // X greater and Y minor. WRO final points
+    initial_y[2] = limit_y - m_resolution;
+    initial_y[3] = limit_y - m_resolution;
+
+    final_y[1] = limit_y - m_resolution;
+    final_y[3] = limit_y - m_resolution;
+
+    min_y = limit_y - m_resolution;
+    max_y = limit_y;
+  }
+
+  if (final_grid_pos[0] < robot_grid_pos[0] && final_grid_pos[1] < robot_grid_pos[1])
+  {
+    // X minor and Y minor. WRO final points
+    initial_x[2] = limit_x - m_resolution;
+    initial_x[3] = limit_x - m_resolution;
+    initial_y[2] = limit_y - m_resolution;
+    initial_y[3] = limit_y - m_resolution;
+
+    final_x[0] = limit_x - m_resolution;
+    final_x[2] = limit_x - m_resolution;
+    final_y[1] = limit_y - m_resolution;
+    final_y[3] = limit_y - m_resolution;
+
+    min_x = limit_x - m_resolution;
+    max_x = limit_x;
+    min_y = limit_y - m_resolution;
+    max_y = limit_y;
+  }
+}
+
+void LifelongSlamToolbox::initializeGrids()
+{
+  /*
+    To create the grid
+  */
+  for (int i = 0; i < m_num_cells; ++i)
+  {
+    // Adding columns
+    m_grid[i].resize(m_num_cells);
+    m_mutual_grid[i].resize(m_num_cells);
+    for (int j = 0; j < m_num_cells; ++j)
+    {
+      m_grid[i][j] = 0;
+      m_mutual_grid[i][j] = 0.0f;
+    }
+  }
+}
+
+float LifelongSlamToolbox::measurementOutcomeEntropy(Occupancy const& meas_outcome)
+{
+  /*
+    To calculate the measurement outcome entropy (Measurement outcome in the form <fr, oc, un>)
+  */
+  float cell_logs = (meas_outcome.fr * calculateLogs(0.3f)) + (meas_outcome.oc * calculateLogs(0.7f)) + (meas_outcome.un * calculateLogs(0.5f));
+  return calculateEntropy(probabilityFromLogs(cell_logs));
+}
+
+float LifelongSlamToolbox::calculateEntropy(float probability)
+{
+  /*
+    To calculate the entropy
+  */
+  return probability * log(probability); 
```

Comment:
@kmilo7204 nit: consider using `log2` instead (to match the usual entropy definition).

Reply:
Noted!